bin/node/src/config.rs

```diff
@@ -1,35 +1,71 @@
+use std::path::Path;
+
 use miden_node_block_producer::config::BlockProducerConfig;
 use miden_node_rpc::config::RpcConfig;
 use miden_node_store::config::StoreConfig;
+use miden_node_utils::config::Endpoint;
 use serde::{Deserialize, Serialize};
 
 /// Node top-level configuration.
-#[derive(Clone, Eq, PartialEq, Ord, PartialOrd, Hash, Debug, Serialize, Deserialize)]
+#[derive(Clone, Default, Eq, PartialEq, Ord, PartialOrd, Hash, Debug, Serialize, Deserialize)]
 pub struct NodeConfig {
-    pub block_producer: Option<BlockProducerConfig>,
-    pub rpc: Option<RpcConfig>,
-    pub store: Option<StoreConfig>,
+    pub block_producer: BlockProducerConfig,
+    pub rpc: RpcConfig,
+    pub store: StoreConfig,
 }
 
-impl Default for NodeConfig {
-    fn default() -> Self {
-        Self {
-            block_producer: Some(Default::default()),
-            rpc: Some(Default::default()),
-            store: Some(Default::default()),
+impl NodeConfig {
+    pub fn load(config_file: impl AsRef<Path>) -> figment::Result<Self> {
+        /// A helper struct to deserialize a subset of [BlockProducerConfig] options.
+        #[derive(Deserialize)]
+        struct UserBlockProducerConfig {
+            endpoint: Endpoint,
+            verify_tx_proofs: bool,
+        }
+
+        /// A helper struct to deserialize a subset of [RpcConfig] options.
+        #[derive(Deserialize)]
+        struct UserRpcConfig {
+            endpoint: Endpoint,
+        }
+
+        /// A normalized version of [NodeConfig] i.e. with all redundant options
+        /// removed. This helps prevent user configuration issues.
+        #[derive(Deserialize)]
+        struct UserConfig {
+            block_producer: UserBlockProducerConfig,
+            rpc: UserRpcConfig,
+            store: StoreConfig,
         }
```

Comment:
nit: I would probably move these structs out of the function definition.

Also, question: how would this work when parts of the config file are missing? For example, say we want to run just the store component of the node (i.e., via `miden-node start store` command). Our config file may look like so:
```toml
[store]
endpoint = { host = "localhost", port = 28943 }
database_filepath = "/opt/miden/miden-store.sqlite3"
genesis_filepath = "/opt/miden/genesis.dat"
blockstore_dir = "/opt/miden/blocks"
```
Similarly, if we wanted to run just the RPC component, the config file may look like so:
```toml
[rpc]
endpoint = { host = "0.0.0.0", port = 57291 }
block_producer_url = "http://block_producer_machine:48046"
store_url = "http://store_machine:28943"
```
I think the above would currently fail to parse, right?

I wonder if we should make the logic a bit more sophisticated. Specifically: if a dependency for a component is present, we try to infer the relevant fields from it or verify that all fields are consistent. To give a few examples (using only block producer and store components):

If the config file has just the `block_producer` component, it could look like so:
```toml
[block_producer]
endpoint = { host = "localhost", port = 48046 }
store_url = "http://localhost:28943"
verify_tx_proofs = true
```

If it has both `block_producer` and `store` components, it could look like so:
```toml
[block_producer]
endpoint = { host = "localhost", port = 48046 }
verify_tx_proofs = true

[store]
endpoint = { host = "localhost", port = 28943 }
database_filepath = "/opt/miden/miden-store.sqlite3"
genesis_filepath = "/opt/miden/genesis.dat"
blockstore_dir = "/opt/miden/blocks"
```
In the above, we infer `block_producer.store_url` from the `store.endpoint`. 

But the following would still be valid:
```toml
[block_producer]
endpoint = { host = "localhost", port = 48046 }
store_url = "http://localhost:28943"
verify_tx_proofs = true

[store]
endpoint = { host = "localhost", port = 28943 }
database_filepath = "/opt/miden/miden-store.sqlite3"
genesis_filepath = "/opt/miden/genesis.dat"
blockstore_dir = "/opt/miden/blocks"
```
This is because `block_producer.store_url` is consistent with the `store.endpoint`.

However, if they were inconsistent (like in the below) - we should get an error:
```toml
[block_producer]
endpoint = { host = "localhost", port = 48046 }
store_url = "http://localhost:12345"
verify_tx_proofs = true

[store]
endpoint = { host = "localhost", port = 28943 }
database_filepath = "/opt/miden/miden-store.sqlite3"
genesis_filepath = "/opt/miden/genesis.dat"
blockstore_dir = "/opt/miden/blocks"
```

The above are just examples because having just 2 sections in a file would not be legal. In my mind, the legal combinations are:
- Either way have all 3 sections present (this is when the entire node runs in a single process), or
- Only one of the sections is present (this is when individual components run in separate processes).

Reply:
Ah I had assumed that running individual components used the individual configurations structs only i.e. that `miden-node start rpc` used `RpcConfig` and not `NodeConfig`.

I'll dig a little deeper into the code and come up with something more elegant :). Moving PR to draft for now.

Reply:
One option is to change the config into an enum to look something like:
```rust
pub enum NodeConfig {
    Full(FullNodeConfig),
    Rpc(RpcConfig),
    BlockProducer(BlockProducerConfig),
    Store(StoreConfig)
}

pub struct FullNodeConfig {
    pub block_producer: BlockProducerConfig,
    pub rpc: RpcConfig,
    pub store: StoreConfig,
}
```
But there are probably other ways too.

Reply:
I went a different route. I handle the different start "modes" configs as separate type so we have full type safety.